GOAT-Compiler/Code Generation/BuildInFunctionImplementations.cs

```diff
@@ -0,0 +1,343 @@
+                    break;
+                case "AbsArcCCW":
+                    AbsArc(actuelParams[0], actuelParams[1], true);
+                    break;
+                case "SetExtrusionRate":
+                    SetExtrusionRate(actuelParams[0]);
+                    break;
+                case "SetBedTemp":
+                    SetBedTemp(actuelParams[0]);
+                    break;
+                case "SetExtruderTemp":
+                    SetExtruderTemp(actuelParams[0]);
+                    break;
+                case "SetFanPower":
+                    SetFanPower(actuelParams[0]);
+                    break;
+                case "Steps":
+                    Steps(actuelParams[0]);
+                    break;
+                case "Position":
+                    return Position();
+                case "Lift":
+                    Lift(actuelParams[0]);
+                    break;
+                case "Right":
+                    Right(actuelParams[0]);
+                    break;
+                case "Left":
+                    Left(actuelParams[0]);
+                    break;
+                case "Direction":
+                    return Direction();
+                case "TurnTo":
+                    TurnTo(actuelParams[0]);
+                    break;
+                case "WaitForBedTemp":
+                    WaitForBedTemp();
+                    break;
+                case "WaitForExtruderTemp":
+                    WaitForExtruderTemp();
+                    break;
+                case "WaitForCurrentMove":
+                    WaitForCurrentMove();
+                    break;
+                case "WaitForMillis":
+                    WaitForMillis(actuelParams[0]);
+                    break;
+                case "Home":
+                    Home();
+                    break;
+            }
+            return null;
+        }
+
+        private void Home()
+        {
+            string gLine = "G28";
+            _machine.Position = new Vector(0, 0, 0);
+            _stream.WriteLine(gLine);
+        }
+
+        private void RelMove(Vector v)
+        {
+            string gLine = "";
+            Vector oldPosition = _machine.Position;
+            _machine.Position = new Vector(oldPosition.X + v.X, oldPosition.Y + v.Y, oldPosition.Z + v.Z);
+            if (_machine.Build == BuildScope.build)
+            {
+                _machine.CurrentExtrusion += (_machine.ExtrusionRate*VectorDistance(oldPosition, _machine.Position));
+                gLine += "G1 X" + (decimal)_machine.Position.X + " Y" + (decimal)_machine.Position.Y + " Z" + (decimal)_machine.Position.Z + " E" + (decimal)_machine.CurrentExtrusion;
+            }
+            else if (_machine.Build == BuildScope.walk)
+            {
+                gLine += "G0 X" + (decimal)_machine.Position.X + " Y" + (decimal)_machine.Position.Y + " Z" + (decimal)_machine.Position.Z;
+            }
+            else
```

Comment:
plz make helper function which takes the _michine.Build and a string of the name of the command, and throws an exception if necessary:
void f(BuildScope machineExtrusionMode, string MovementFunctionName){
if (machineExtrusionMode == BuildScope.none)
 MoveWithoutScopeException($"{MovementFunctionName} cant be called without being in a build or walk scope.");
}

Reply:
Added

Reply:
fixed